lib/portage/package/ebuild/doebuild.py

```diff
@@ -2133,9 +2187,65 @@ def spawn(
         logname_backup = mysettings.configdict["env"].get("LOGNAME")
         mysettings.configdict["env"]["LOGNAME"] = logname
 
+    eapi = mysettings["EAPI"]
+
+    unexported_env_vars = None
+    if "export-pms-vars" not in mysettings.features or not eapi_exports_pms_vars(eapi):
+        unexported_env_vars = _unexported_pms_vars
+
+    if unexported_env_vars:
+        # Starting with EAPI 9 (or if FEATURES="-export-pms-vars"),
+        # PMS variables should not longer be exported.
+
+        phase = mysettings.get("EBUILD_PHASE")
+        is_pms_ebuild_phase = phase in _phase_func_map.keys()
+        # 'None' phase is MiscFunctionsProcess, e.g., where the qa checks run
+        is_ebuild_phase_with_t = phase in [None, "package"]
+        # Copy the environment since we are removing the PMS variables from it.
+        env = mysettings.environ().copy()
+        if is_pms_ebuild_phase or is_ebuild_phase_with_t:
+            t = env["T"]
+            ebuild_extra_source_path = os.path.join(
+                t, f".portage-ebuild-extra-source-{phase}"
+            )
+        else:
+            global _emerge_tmpdir
+            if _emerge_tmpdir is None:
+                _emerge_tmpdir = tempfile.mkdtemp(
+                    prefix=f"portage-tmpdir-{portage.getpid()}-"
+                )
```

Comment:
This parent directory needs to be made world executable.

Reply:
This doesn't seem to be required. The phases that trigger this case seem to be able to access it. Which makes me wonder if, due the latest changes, the chmod() above is still required.

However, I can see an argument to make it always world readable. That would also helpwith debugging. In contrast to the argument to not introduce unnecessary code.

So we probably should either chmod the file and the directory, or do none of that.

I am currently leaning towards making it world readable, but first let me run some experiments.

Reply:
Also, the atexit hook is not ideal for things like egencache that will create a large number of temporary files. It would be much nicer to have EbuildMetadataPhase clean up the temporary files as soon as their lifecycle is complete.

Reply:
Since ebuild.sh removes the up the file, the at exit hook just cleans up the (empty) directory.

I just shove all files under a directory to avoid cluttering the tmp directory.

Reply:
The `_emerge_tmpdir` dir is now world-readable.

Reply:
Removing the file in ebuild.sh will work, but the parent directory needs to be writable by the portage user.

Reply:
I don't see any current issues with the directory not being writable by the portage user. I assume all phases that hit the case are running as root. To make sure, I've added a `|| die` to the rm in the latest version of the patch.

Are you sure that there is an issue here?

Reply:
The depend phase should not run as root if userpriv is enabled. You can run egencache to trigger it.

Reply:
Alright, I will look into it. Thanks.

I wonder if we should set the sticky bit as well, as otherwise unprivileged phases could delete files from privileged ones.

Reply:
I gave it a spin, and the trap never executes because the depend phase never calls __ebuild_main.

Reply:
Got it, thanks!

So we need to consider three cases: A) ebuild phase with T available. B) privileged ebuild phase with T unavailable and C) phase=depend

For C, I have resurrected the life cycle managed via `EbuildMetadataPhase`. Please have a look at the latest revision of the patch.

And thanks for your assistance and patience. That was a tremendous help.